cmd/src/gitserve.go

```diff
@@ -0,0 +1,370 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+	"text/template"
+
+	"github.com/pkg/errors"
+)
+
+func init() {
+	flagSet := flag.NewFlagSet("git-serve", flag.ExitOnError)
+	usageFunc := func() {
+		fmt.Fprintf(flag.CommandLine.Output(), `'src git-serve' is a tool which serves your local git repositories over HTTP for Sourcegraph to pull.
+
+USAGE
+  src [-v] git-serve [-addr :3434] [path/to/dir]
+
+By default 'src git-serve' will recursively serve your current directory on the address ':3434'.
+`)
+	}
+	var (
+		addrFlag = flagSet.String("addr", ":3434", "Address on which to serve (end with : for unused port)")
+	)
+
+	handler := func(args []string) error {
+		err := flagSet.Parse(args)
+		if err != nil {
+			return err
+		}
+
+		var repoDir string
+		switch args := flagSet.Args(); len(args) {
+		case 0:
+			repoDir, err = os.Getwd()
+			if err != nil {
+				return err
+			}
+
+		case 1:
+			repoDir = args[0]
+
+		default:
+			return &usageError{errors.New("requires zero or one arguments")}
+		}
+
+		dbug := log.New(ioutil.Discard, "", log.LstdFlags)
+		if *verbose {
+			dbug = log.New(os.Stderr, "DBUG git-serve: ", log.LstdFlags)
+		}
+
+		s := &Serve{
+			Addr:  *addrFlag,
+			Root:  repoDir,
+			Info:  log.New(os.Stderr, "git-serve: ", log.LstdFlags),
+			Debug: dbug,
+		}
+		return s.Start()
+	}
+
+	// Register the command.
+	commands = append(commands, &command{
+		flagSet:   flagSet,
+		handler:   handler,
+		usageFunc: usageFunc,
+	})
+}
+
+type Serve struct {
+	Addr  string
+	Root  string
+	Info  *log.Logger
+	Debug *log.Logger
+
+	// updatingServerInfo is used to ensure we only have 1 goroutine running
+	// git update-server-info.
+	updatingServerInfo uint64
+}
+
+func (s *Serve) Start() error {
+	ln, err := net.Listen("tcp", s.Addr)
+	if err != nil {
+		return errors.Wrap(err, "listen")
+	}
+
+	// Update Addr to what listener actually used.
+	s.Addr = ln.Addr().String()
+
+	s.Info.Printf("listening on http://%s", s.Addr)
+	h, err := s.handler()
+	if err != nil {
+		return errors.Wrap(err, "configuring server")
+	}
+
+	if err := (&http.Server{Handler: h}).Serve(ln); err != nil {
+		return errors.Wrap(err, "serving")
+	}
+
+	return nil
+}
+
+var indexHTML = template.Must(template.New("").Parse(`<html>
+<head><title>src git-serve</title></head>
+<body>
+<h2>src git-serve</h2>
+<pre>
+{{.Explain}}
+<ul>{{range .Links}}
+<li><a href="{{.}}">{{.}}</a></li>
+{{- end}}
+</ul>
+</pre>
+</body>
+</html>`))
+
+type Repo struct {
+	Name string
+	URI  string
+}
+
+func (s *Serve) handler() (http.Handler, error) {
+	s.Info.Printf("serving git repositories from %s", s.Root)
+	s.configureRepos()
```

Comment:
I'd pull these two out of `handler()` and put them in `Start()`, where the rest of the imperative code lives. That would make `handler()` a "pure" method that only returns the handler.

btw. the `s.configureRepos()` is also in `mux.HandleFunc("/v1/list-repos", func(w http.ResponseWriter, r *http.Request) { ... }` — the call here is only done for side-effects I assume?

Reply:
will do. Yes it is done for side-effect. It's a side-effect we can remove in another commit I am getting done as part of making src-expose better.

Reply:
we don't actually need the side-effect anymore, so nice spotting.